kobotoolbox/kpi
# feat(projectHistoryLogs): remove audit logs according to constance configuration TASK-972

## Checklist

1. [x] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Run `./python-format.sh` to make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/main/CONTRIBUTING.md)
5. [x] Write a title and, if necessary, a description of your work suitable for publishing in our [release notes](https://community.kobotoolbox.org/tag/release-notes)
6. [x] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes
8. [x] Create a testing plan for the reviewer and add it to the Testing section
9. [x] Add frontend or backend tag and any other appropriate tags to this pull request

## Description

This PR adds the automatic expiration mechanism for project history logs. It is similar to the access log mechanism. The expiration time span is configured through constance.

## Notes

All changes are on the audit_log kobo app. 

## Testing

1. Change the title of an asset on Kobo
2. Open a shell on django
```
$ ./run.py -cf exec kpi bash 
$ ./manage.py shell_plus
```
3. Check that the change you just made is reflected in the last log metadata:
```
In [7]: ProjectHistoryLog.objects.last().metadata
Out[7]: 
{'name': {'new': 'Test test log', 'old': 'Test'},
 'source': 'Chrome (Linux)',
 'asset_uid': 'ae6hQt9s2phfjRnsNphLdA',
 'ip_address': '172.25.0.1',
 'log_subtype': 'project',
 'latest_version_uid': 'vi7x8mXHoWg7SVuW8sCnCS'}
``` 
3. On the Django admin screen, go to constance Config and set the value of PROJECT_HISTORY_LOG_LIFESPAN to `0` 
4. Execute the cleaning task:

```
In [8]: from kobo.apps.audit_log.tasks import spawn_logs_cleaning_tasks

In [9]: spawn_logs_cleaning_tasks()
```
5. Check that there are no project history logs:
```
In [11]: len(ProjectHistoryLog.objects.all())
Out[11]: 0
```

## Related issues

Fixes Notion TASK-972
